### PR TITLE
dplyr specific data pronoun

### LINKED ISSUE #3814: Stronger `.data` pronoun
With the redesign of the standard evaluation path in #3799 we need a `.data` pronoun that looks in an ancestry of environments. 

We can incubate this here, but I guess this is more something that should live in `rlang`

### PR DESCRIPTION
**Pull request by romainfrancois**
To accommodate for dplyr's data mask using two environments: resolved that masks active which contains all the active bindings. 

closes #3814.

### CODE REVIEW

**Comment by romainfrancois** [comment]
Initially I wanted this to be more flexible with a `while` loop looking between the `.top_env` until the `.env` but I wasn't sure then how to implement `names` and `length` so this version is fine for `dplyr`. 

I still think a data pronoun constructor that takes just a data mask would make sense in `rlang`

File: src/data_pronoun.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+// simple dplyr specific data pronoun that needs to look in two environments
+// stored in a list
+SEXP data_pronoun(SEXP data_mask) {
+  SEXP pronoun = PROTECT(Rf_allocVector(VECSXP, 2));
+  SEXP resolved = Rf_findVarInFrame3(data_mask, Rf_install(".top_env"), FALSE);
+  SEXP active = ENCLOS(resolved);
+  SET_VECTOR_ELT(pronoun, 0, resolved);
+  SET_VECTOR_ELT(pronoun, 1, active);
+  Rf_setAttrib(pronoun, R_ClassSymbol, Rf_mkString("dplyr_data_pronoun"));
```
Comment (lionel-) [inline src/data_pronoun.cpp]: `mkString()` needs a protect as setting attributes allocates cons cells.

File: src/data_pronoun.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#include "pch.h"
+#include <dplyr/main.h>
+
+// simple dplyr specific data pronoun that needs to look in two environments
+// stored in a list
+SEXP data_pronoun(SEXP data_mask) {
+  SEXP pronoun = PROTECT(Rf_allocVector(VECSXP, 2));
+  SEXP resolved = Rf_findVarInFrame3(data_mask, Rf_install(".top_env"), FALSE);
+  SEXP active = ENCLOS(resolved);
```
Comment (lionel-) [inline src/data_pronoun.cpp]: I think this might crash if the top_env pronoun is not found.

**Comment by lionel-** [comment]
> On second thought I guess that's more prudent to use custom classes. For instance tidyselect masks c() in a higher-level environment, we probably don't want .data$c to return that function.

Maybe we could make `$` look up in all environments of the mask marked with a specific flag? Let's keep these issues in mind, maybe more cases will come up.

File: src/data_pronoun.cpp (hunk 1 above)
Comment (romainfrancois) [reply src/data_pronoun.cpp]: Thanks.

File: src/data_pronoun.cpp (hunk 2 above)
Comment (romainfrancois) [reply src/data_pronoun.cpp]: I'm only ever calling this right after calling `rlang::new_data_mask`: 

```
      data_mask = rlang::new_data_mask(
                    mask_resolved, // bottom
                    mask_active    // top
                  );

      // install the pronoun
      Rf_defineVar(symbols::dot_data, data_pronoun(data_mask), data_mask);
```

And this `data_pronoun` function is not exported, it's only meant for dplyr internals. So I'd say that's fine.

**Comment by lionel-** [comment]
In fact I think we might want to disallow `names()`, `length()` and `[`. The data pronoun has no notion of ordering to start with (a user opened an issue about this as this was confusing).

File: src/data_pronoun.cpp (hunk 2 above)
Comment (lionel-) [reply src/data_pronoun.cpp]: I guess we should lock the special mask bindings.

**Comment by romainfrancois** [comment]
Right ok, in that case, I'll just open an issue in rlang and we can resume the conversation there.

**Comment by lionel-** [comment]
Issue where this was discussed: https://github.com/tidyverse/dplyr/issues/3734

@hadley confirmed there the contract should only be about `$` and `[[`.

**Comment by romainfrancois** [comment]
`length()` is used by the `print()` method though

**Comment by lionel-** [comment]
I think it's ok to just print `<pronoun>` since it's not an object that is straightforward to print. We have to insert a `browser()` or a `foo <<- .data` in dplyr calls to print it.

**Comment by lionel-** [comment]
Maybe we can offer some advice about using the pronoun in the print method for people trying `rlang::.data` at the console.

**Comment by romainfrancois** [comment]
superseded by #3880 which uses something more generic from `rlang`.

**Pull request closed without merging**